tests/server/job/test_job.py

```diff
@@ -0,0 +1,322 @@
+    then return 100 on the `succeed_on_nth_run` run. The `tmp_dir` records the run state.
+    """
+    from mlflow.server.job import TransientError
+
+    if len(os.listdir(tmp_dir)) == succeed_on_nth_run:
+        return 100
+    with open(os.path.join(tmp_dir, uuid.uuid4().hex), "w") as f:
+        f.close()
+    raise TransientError(RuntimeError("test transient error."))
+
+
+def test_job_retry_on_transient_error(monkeypatch, tmp_path):
+    monkeypatch.setenv("MLFLOW_SERVER_JOB_TRANSIENT_ERROR_RETRY_BASE_DELAY", "1")
+    with _setup_job_runner(1, monkeypatch, tmp_path):
+        store = _get_job_store()
+
+        job1_tmp_path = tmp_path / "job1"
+        job1_tmp_path.mkdir()
+
+        job1_id = submit_job(
+            transient_err_fun, {"tmp_dir": str(job1_tmp_path), "succeed_on_nth_run": 4}
+        )
+        time.sleep(15)
+        assert query_job(job1_id) == (JobStatus.FAILED, "RuntimeError('test transient error.')")
+        job1 = store.get_job(job1_id)
+        assert job1.status == JobStatus.FAILED
+        assert job1.result == "RuntimeError('test transient error.')"
+        assert job1.retry_count == 3
+
+        job2_tmp_path = tmp_path / "job2"
+        job2_tmp_path.mkdir()
+
+        job2_id = submit_job(transient_err_fun, {"tmp_dir": str(job2_tmp_path), "succeed_on_nth_run": 1})
+        time.sleep(3)
+        assert query_job(job2_id) == (JobStatus.DONE, 100)
+        job2 = store.get_job(job2_id)
+        assert job2.status == JobStatus.DONE
+        assert job2.result == "100"
+        assert job2.retry_count == 1
+
+
+# `submit_job` API is designed to be called inside MLflow server handler,
+# MLflow server handler might be executed in multiple MLflow server workers
+# so that we need a test to cover the case that executes `submit_job` in
+# multi-processes case.
+def test_submit_jobs_from_multi_processes(monkeypatch, tmp_path):
+    with _setup_job_runner(4, monkeypatch, tmp_path), MultiProcPool() as pool:
+        async_res_list = [
+            pool.apply_async(
+                submit_job,
+                args=(basic_job_fun,),
+                kwds={"params": {"x": x, "y": 1, "sleep_secs": 2}},
+            )
+            for x in range(4)
+        ]
+        job_ids = [async_res.get() for async_res in async_res_list]
+        time.sleep(3)
+        for x in range(4):
+            assert query_job(job_ids[x]) == (JobStatus.DONE, x + 1)
+
+
+def sleep_fun(sleep_secs, tmp_dir):
+    (Path(tmp_dir) / "pid").write_text(str(os.getpid()))
+    time.sleep(sleep_secs)
+
+
+def test_job_timeout(monkeypatch, tmp_path):
+    from mlflow.server.job.util import is_process_alive
+
+    with _setup_job_runner(1, monkeypatch, tmp_path) as job_runner_proc:
+        job_tmp_path = tmp_path / "job"
+        job_tmp_path.mkdir()
+        job_id = submit_job(
+            sleep_fun, {"sleep_secs": 10, "tmp_dir": str(job_tmp_path)}, timeout=5
```

Comment:
do we really need to sleep for 10 seconds?

Reply:
`sleep_secs` does not matter, the timeout `5` matters here.

I use timeout 5s (not shorter timeout) for avoid flakiness.

In my local test, launching the sub-process needs 2~3 seconds (because it uses `spawn` mode instead of `fork` mode, so that it takes more time)